### DISCUSSION TITLE
Future features for Notification interface

### DISCUSSION THREAD

**Discussion by jsparber** [flatpak/xdg-desktop-portal#1495] [New Portals]
A lot of features proposed in https://github.com/flatpak/xdg-desktop-portal/issues/983 landed now in version 2. Some features didn't make the cut. This tries to collect additional features we discussed already while creating v2. Some of them would be nice to have in near future for v3.

- Threading
  Group notifications from the same app into a thread, eg. for multiple chat messages from one contact

- Vibration
  Allow an app to set a vibration pattern played when the notification is shown. This can be based on the [Web Vibration API specification](https://developer.mozilla.org/en-US/docs/Web/API/Vibration_API) which is similar to [android](https://developer.android.com/reference/android/app/NotificationChannel#setVibrationPattern(long[]))'s API

- Notification Led
  Allow apps to specify a led color that will be used on devices that have a notification led. [Android](https://developer.android.com/reference/android/app/NotificationChannel#setLightColor(int)) has an API for it. 

- Progress indicator
  Show a progress bar in notifications e.g for file copy
 
- Image preview
  Large image for the body of the notification, including alt text for accessibility

- Alternative notification
   Allow apps to specify an alternative notification for the lockscreen
   
- Timeout
 Allow apps to specify I timeout of timestamp when the event ends, useful for things like calendar appointments

- `desktop-file-id`
  Application settable desktop file id.
  Partially solved with https://github.com/flatpak/xdg-desktop-portal/pull/1488

- Permission requests
  Ask the user whether they want to receive notifications from this app. Ideally, apps should ask for permission before sending a notification, since the permission dialog shouldn't popup at random but when the user expects it.

Feel free to comment if I missed features we discussed already and didn't land with v2.

**Comment by Mikenux** [comment]
I changed my mind about the permission to send notifications, but I take that as a "reason" to run in the background.

**Reply by Mikenux** [reply]
I opened a discussion: https://github.com/flatpak/xdg-desktop-portal/discussions/1496.

Note that the second user experience (when the app is closed) is to let the app send in-app notifications. For example, this is (or was?) the case of Dolphin, the KDE file manager, which displays the progress of file operations.

**Comment by DemiMarie** [comment]
What about:

- Persistent notifications that persist after the app closes.
- Feature parity with macOS and Windows.

**Reply by jsparber** [reply]
> Persistent notifications that persist after the app closes.

Already the case, if the notification server implements it and the app uses D-Bus Activation.

> Feature parity with macOS and Windows.

Would love that, but not relevant for xdg portals since they aren't used on Mac and Windows


The above list wasn't intended to be an exhaustive list, but more a collection of features that were discussed already at some point.

**Reply by DemiMarie** [reply]
> > Persistent notifications that persist after the app closes.
> 
> Already the case, if the notification server implements it and the app uses D-Bus Activation.

That is great!  I thought this was not supported, but I am glad to be wrong.  Is there some sort of ID that can be used to identify the notification, or should the app just include that itself?

> > Feature parity with macOS and Windows.
> 
> Would love that, but not relevant for xdg portals since they aren't used on Mac and Windows

I meant “feature parity with macOS and Windows notification APIs”, meaning that notifications are as featureful on Linux as they are on proprietary desktop OSs in the absence of a good reason (such as security) for them not to be.

**Comment by Levi-Lesches** [comment]
Hi @jsparber @jonas2515, I was reading through the v2 proposal and saw this: 

> **Things we don't want**
> 
> - We don't want API for showing "n unseen notifications" badges on app icons.

I didn't see this discussed anywhere in that thread -- but why not? Badges are supported on every other platform: [Windows](https://learn.microsoft.com/en-us/uwp/api/windows.ui.notifications.badgeupdatemanager?view=winrt-26100), [MacOS, iOS](https://developer.apple.com/documentation/usernotifications/unnotificationcontent/badge), [Android](https://developer.android.com/develop/ui/views/notifications/badges#set-custom-count), and even [Web](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Display_badge_on_app_icon). Googling "Linux app badges" shows bug reports (mostly about Discord) from users concerned that they can't see their badges, so I'd argue it's a feature people want. Is this already implemented/described elsewhere, or is this really not planned?

**Reply by orowith2os** [reply]
One could argue that this is up to the shell to deal with - it knows what notifications are seen or unseen, and where they came from. It should be able to match them up appropriately, I'd think?

**Reply by orowith2os** [reply]
Oh, never mind, I see what you're talking about. Yeah, that makes sense. Just passing a number for unread messages would make sense, I think. Any more information on this?

**Reply by Mikenux** [reply]
I think a dedicated discussion should be open about this.

**Reply by Levi-Lesches** [reply]
This thread seems general enough, as it talks about future features in general. 

@jsparber, if you think it's appropriate, I can open a new thread for this.

**Reply by Mikenux** [reply]
Note: This could perhaps work as a category of new notifications. This means that an "unread" badge will be limited to **new** unread messages, not those that exist before an app is installed.

**Reply by Levi-Lesches** [reply]
@Mikenux given it's been two months and there is other activity on the repo, I've opened a new discussion at #1617

**Comment by 12people** [comment]
I'd love to see:
- Scheduled notifications
- Repeat notifications

(Both of those are available on Android, iOS, and macOS.)

I started a separate conversation on this topic: #1696

**Comment by Mikenux** [comment]
Badges: https://github.com/flatpak/xdg-desktop-portal/discussions/1702. Having something specific for badges wasn't avoidable. However, this does not prevent a server from not supporting them.

**Comment by Mikenux** [comment]
Awareness of Do Not Disturb: https://github.com/flatpak/xdg-desktop-portal/discussions/1321

**Comment by jadahl** [comment]
> Progress indicator
> Show a progress bar in notifications e.g for file copy

I wonder if this could be morphed into the use case of showing progress bars on dash/dock app icons without an actual notification bubble, or if that should be handled separately.

**Reply by Mikenux** [reply]
Just let the system decide how and where to draw that?

**Reply by jadahl** [reply]
Well, that is the question, to some degree. Should there be enough metadata for the system to make such a decision?

**Reply by Mikenux** [reply]
Sorry, mistake. Replied too quickly.

It depends on which progress.

For files, separately seems better, maybe in the file chooser portal (save part) with relevant properties, and that’s the system that decides to report it.

**Reply by jadahl** [reply]
Doesn't have to do anything with files, maybe it's just a "Calculate Pi" app that has a progress bar of calculating the Nth decimal of Pi, or anything else non-file-related.

**Reply by Mikenux** [reply]
I mean any propress non-related to files here. Alternatively, this might also be an app status (status notifier), which also needs to be handled: Is it something separate? Do app devs expect to report an app status and a progress at the same time? (So, here, I have a part of the response on the notification badges discussion).

**Reply by razzeee** [reply]
I'm now wondering about visually distinguishing failed progress from successful finished progress

**Reply by Play-InTheWay** [reply]
> I'm now wondering about visually distinguishing failed progress from successful finished progress

Normally it is done through coloring the progress bar with a specific set of colors, mainly being

- Green: in progress
- Yellow: Paused
- Red: Error/Stopped

The themes question shouldn't be an issue here, because as far as I know, they also have specified colors for similar things in the interface with names like “success, warning, or error.” So they should be used there as well.

**Reply by razzeee** [reply]
Which wouldn't be accessible to Protanopia people or similar. I also specifically asked for the end state - as in progress is pretty easy to understand (but I guess adding a pause complicates things).

**Reply by ssokolow** [reply]
Does the Windows equivalent distinguish them?

From what I remember from last time I used Windows 7, there's just progress and, if user attention is needed, it sets the "needs attention" flag on the taskbar button  and displays the completion status in the window.

**Reply by Mikenux** [reply]
For a DE to properly consider accessibility for progresses, those need to be clearly specified in a portal: their types and different states. How the information is visually presented is up to the DE.

**Reply by AdrianVovk** [reply]
I like this idea. I mean we could do both, show a notification w/ a progress bar (or multiple notifications w/ progress bars) and then on the app's icon show "overall progress". There was a [GNOME Design Whiteboard](https://gitlab.gnome.org/Teams/Design/whiteboards/-/work_items/274) about having a status icon with a drop down as well to show "system-wide progress".

In general, we should have bespoke UI for certain types of notifications where it makes sense. Progress notifications are just one type that we may find interesting. Ringing alarms, or incoming;ongoing calls, etc we may also want to special case with special UI. Also things like navigation turn-by-turn directions, upcoming travel, etc. For instance, both Android and iOS show a little summary of these kinds of notifications in the status bar and have other places in the system UI where they get special integration
